### ISSUE TITLE
Disable cache file looses history between Gnome lock and unlock

### ISSUE THREAD

**Issue by JPvRiel** [Tudmotu/gnome-shell-extension-clipboard-indicator#91]
On ubuntu 16.04.3 LTS, it's very undesirable to loose history between locking and unlocking the session.

For whatever reason, it seems Gnome is unloading/reloading the extensions between lock and unlock and this implies the approach taken with disable cache has this fairly painful limitation. Fixing it may require a major refactoring and further investigation. It's only worth the effort if enough users want this feature...

**Comment by freeroot** [comment]
I hope I don't say anything wrong.

In your last comment of #58, you told about to guarantee security of items in clipboard indicator to be not accessible to other running programs. So, don't use tmpfs or some things like that.
I don't understand why it is so important to have this guarantee for "disable cache file" option whereas, without "disable cache file" option enabled, there is a registry.txt file in ~/.cache and there is not security guarantee.

So, in your opinion, "disable cache file" is a security option or a enhancement option ?
If it's a security option, let's talk about registry.txt too and, logically, we should open a specific issue for those questions.
If it's a enhancement option, tmpfs or other ideas shouldn't be rejected.

**Comment by JPvRiel** [comment]
So I'm thinking you want a button/option to move cache/registry.txt to tmpfs (.e.g. "non-persistent cache", or custom location), and that can be an independent option to the one I added called "disable cache".

> enhancement option, tmpfs

Your idea is valid (If I assume you mean "don't keep my clipboard history between user sessions") and then using tmpfs will help. E.g. perhaps look at implementing an advanced option to allow users to configure where registry.txt is placed? Ot that could be somewhere like `/run/user/<uid>/clipboard-indicator@tudmotu.com/registry.txt`. However, note that Gnome might run on more than just Linux (e.g. FreeBSD which doesn't have systemd, etc, and I'm unsure if `/run/user/<uid>/` is supported by all systems running Gnome). So one should rather not hardcode OS specific file paths into. An option to specify a custom registry.txt location can then be up to the few users who care to do it.

The tmpfs option might work well ultimately if/when Wayland / Gnome security improves to put extensions into protected sandboxes.

For a tmpfs workaround, see https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/issues/46#issuecomment-302878531 is as far as I went.

However, if all you want is to delete history between reboots/Gnome session closing, perhaps there's a hook in the GNOME API to tell an extension the session is going to end an allow it a short amount of time to take an action?

> guarantee security of items in clipboard indicator to be not accessible to other running programs

Some of my concern is X11 (in)security. Applications all run under one user account, and can access each other freely. Install and run just one trojaned/malicious application or script and it could read your tmpfs registry.txt clipboard file, etc. By comparison, Android and iOS put a far more restrictive security model in place to sandbox applications (which is why you often have to "share" / transfer data between apps). For the Linux desktop, maybe Wayland (Likely X11 successor) will address some issues but I haven't checked (yet). I wayland could address concerns in future, then maybe a tmpfs approach could work, but it would also likely depend on Gnome's security model as well, and if it bothered to create boundaries/isolation between extensions.

Maybe we shouldn't be expending too much effort to guard one extension when, upstream projects like Wayland / Gnome should be improving the security architecture? Basically, why go to tons of effort to fix it just for this extension? Better economies of scale if Gnome / Wayland projects address some security requirements for ALL extensions?

> So, in your opinion, "disable cache file" is a security option 

I implemented as a work-around to security concerns, so it's primarily a security option (with some bad side-effects).

> I hope I don't say anything wrong.

Being wrong or write isn't a worry, happy to discuss requirements. We probably just have different use cases, security versus stale/old history... :-)

**Comment by freeroot** [comment]
I would have said false instead of wrong because I'm new in security approach but thanks for your conclusion.

Ok, I was thinking about an enhancement option.

I'm afraid that how save datas in security and how don't keep history across reboot couln't be possible simultaneously for now. Maybe in context of insecure Wayland / insecure Gnome, it were not possible. Maybe improve those programs will be the only way. You seem to have good knowledge to find a solution with other people.

In the meantime, neither good security for cache file nor good enhancement because history is deleted by lock/unlock.

In the meantime, deleting cache file when "Disable cache file" option is selected (https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/issues/89) were just to fix a minor bug which is totally weird in every case.

**Comment by JPvRiel** [comment]
> I'm afraid that how save datas in security and how don't keep history across reboot couln't be possible simultaneously for now.

Well, to split our use cases out more neatly:

>  if all you want is to delete history between reboots/Gnome session closing, perhaps there's a hook in the GNOME API to tell an extension the session is going to end an allow it a short amount of time to take an action?

But not sure if the API has that.

Another idea is to leverage the Gnome keyring perhaps, and I'm not sure, but maybe it controls which applications can access which passwords/keys. Given than, one could create a key for clipboard indicator, and then encrypt the registry/cache files. Best of both security and persistent history?

The [keyman](https://extensions.gnome.org/extension/819/keyman/) extension might provide clues on how to do this. No time to research in detail, but this might be the way to go if Wayland and Gnome 3 don't provide clean / isolated paths for persistent extension data.

**Comment by freeroot** [comment]
I'm thinking about encryption too.

The keyman extension shows that is possible to save encryption key in gnome keying. It shows that at least other extensions (maybe other programs) can have access to this key. If Gnome keying can be forced to allow access of a couple password/key only to clipboard indicator, so I think we get it.

**Comment by freeroot** [comment]
I think that we have the answer : https://bugzilla.gnome.org/show_bug.cgi?id=574315#c4 .

Maybe my remark wasn't justified after all^^

**Comment by szpak** [comment]
I've bumped this issue and the extension at all looking for some ClipIt replacement working with Wayland. In general, I like the way it is keyboard oriented and only the problem with loosing history on (un)lock with in-memory model is somehow problematic.

However, I asked complained about that in the [thread](https://discourse.gnome.org/t/multi-element-clipboard-manager-for-gnome-on-wayland/4584/6?u=szpak) where the extension was suggested for the first time and I've got interesting feedback:

```
// You could use a Map, Object, Set or Array; whichever suits the use-case best
const ClipboardHistory = new Map();

function init() {
}

function enable() {
    // Setup for extension

    // Add/remove item history - NOTE: In our case it would be added by intercepting copy to clipboard, not in `enabled()`
    ClipboardHistory.set('history-item-id', 'clipboard content');
    ClipboardHistory.delete('history-item-id');
}

function disable() {
    // Teardown for extension
}
```

I'm not a JavaScript not Gnome extensions developer, but using a nice [video tutorial](http://www.youtube.com/watch?v=iMyR5lJf7dU) (also suggested in that thread) I was able to do a small PoC with keeping the extension state in a map - still in-memory - between across locks and unlocks (also with Wayland).

I wonder, if it couldn't be used in gnome-shell-extension-clipboard-indicator to keep the in-memory entries?

The other thoughts:
1. I haven't seen a `destroy()` method, so accumulated data would be kept in-memory even after the extension was disabled (up until the gnome-shell is restarted). However, gnome-shell-extension-clipboard-indicator seems to limit the size of stored data, so probably it would be rather hard to eat all available memory. Nevertheless, it would be good to mention that in the documentation, as a side effect.
2. As I don't see a practical use case for doing otherwise, it would be probably best to keep pinned/favorites items persistent by default (and keep only non-pinned in-memory).

WDYT?
